src/librustc_codegen_llvm/back/write.rs

```diff
@@ -547,7 +547,13 @@ unsafe fn optimize(cgcx: &CodegenContext,
             llvm::LLVMRustAddAnalysisPasses(tm, fpm, llmod);
             llvm::LLVMRustAddAnalysisPasses(tm, mpm, llmod);
             let opt_level = config.opt_level.unwrap_or(llvm::CodeGenOptLevel::None);
-            let prepare_for_thin_lto = cgcx.lto == Lto::Thin || cgcx.lto == Lto::ThinLocal;
+            // Some options cause LLVM bitcode to be emitted, which uses ThinLTOBuffers, so we need
+            // to make sure we prepare for thin LTO when emitting bitcode; otherwise we'll get
+            // errors in LLVM.
+            let using_thin_buffers = config.emit_bc || config.obj_is_bitcode
+                || config.emit_bc_compressed || config.embed_bitcode;
+            let prepare_for_thin_lto = cgcx.lto == Lto::Thin || cgcx.lto == Lto::ThinLocal
+                || using_thin_buffers;
```

Comment:
Note that PrepareForThinLTO also disables various unrolling / vectorization passes. That's desired if ThinLTO is actually used, but I suspect it's not what you want if we're just embedding bitcode.

Reply:
It's probably sufficient for this case to just run the NameAnonGlobals pass (the unnamed globals are what's causing problems with the thin buffers).

Reply:
That should indeed be sufficient. I'm wondering though why bytecode emission is still done using ThinLTOBitcodeWriter, even if ThinLTO is not used.